File: OneToOneCallSample/app/src/test/java/com/stringee/apptoappcallsample/stringee/integration/CallAndroidIntegrationTest.java
package com.stringee.apptoappcallsample.stringee.integration;

import static android.content.pm.ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION;
import static android.content.pm.ServiceInfo.FOREGROUND_SERVICE_TYPE_MICROPHONE;
import static android.content.pm.ServiceInfo.FOREGROUND_SERVICE_TYPE_PHONE_CALL;
import static android.content.pm.ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.robolectric.Shadows.shadowOf;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.PackageInfo;
import android.content.pm.ServiceInfo;

import androidx.test.core.app.ApplicationProvider;

import com.stringee.apptoappcallsample.stringee.common.CallConstants;
import com.stringee.apptoappcallsample.stringee.common.CallNotificationManager;
import com.stringee.apptoappcallsample.activity.MainActivity;
import com.stringee.apptoappcallsample.stringee.activity.CallActivity;
import com.stringee.apptoappcallsample.stringee.manager.ActivePushCall;
import com.stringee.apptoappcallsample.stringee.service.InCallService;
import com.stringee.apptoappcallsample.stringee.service.IncomingCallService;
import com.stringee.apptoappcallsample.stringee.service.MediaProjectionCallService;

import org.junit.After;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = 35)
public class CallAndroidIntegrationTest {
    private final Context context = ApplicationProvider.getApplicationContext();

    @After
    public void tearDown() {
        ActivePushCall.clear();
    }

    @Test
    public void pushIncomingNotificationTargetsMainActivityAndCarriesPushOwnership() {
        assertTrue(ActivePushCall.claim("call-1"));
        long generation = ActivePushCall.getGeneration();

        Notification notification = CallNotificationManager.getInstance(context)
                .buildIncomingFromPush("alice", "Alice", true, "call-1", generation);

        assertNotNull(notification.contentIntent);
        assertNotNull(notification.actions);
        assertEquals(2, notification.actions.length);
        boolean foundAnswer = false;
        boolean foundReject = false;
        for (Notification.Action action : notification.actions) {
            PendingIntent pendingIntent = action.actionIntent;
            Intent savedIntent = shadowOf(pendingIntent).getSavedIntent();
            if (CallConstants.ACTION_REJECT.equals(savedIntent.getAction())) {
                foundReject = ActivePushCall.owns(
                        savedIntent.getStringExtra(CallConstants.EXTRA_CALL_ID),
                        savedIntent.getLongExtra(CallConstants.EXTRA_CALL_GENERATION, -1));
            }
            if (ActivePushCall.PendingAction.ANSWER.name().equals(
                    savedIntent.getStringExtra(CallConstants.EXTRA_PENDING_ACTION))) {
                foundAnswer = ActivePushCall.owns(
                        savedIntent.getStringExtra(CallConstants.EXTRA_CALL_ID),
                        savedIntent.getLongExtra(CallConstants.EXTRA_CALL_GENERATION, -1));
            }
        }
        assertTrue(foundAnswer);
        assertTrue(foundReject);
        Intent contentIntent = shadowOf(notification.contentIntent).getSavedIntent();
        assertEquals(MainActivity.class.getName(),
                contentIntent.getComponent().getClassName());
    }

    @Test
    public void sdkIncomingNotificationTargetsCallActivityAndCarriesBothOwners() {
        assertTrue(ActivePushCall.claim("call-2"));
        long pushGeneration = ActivePushCall.getGeneration();
        long sessionGeneration = 81;

        Notification notification = CallNotificationManager.getInstance(context)
                .buildIncomingFromSdk("bob", "Bob", false, "call-2",
                        pushGeneration, sessionGeneration);

        Intent contentIntent = shadowOf(notification.contentIntent).getSavedIntent();
        assertEquals(CallActivity.class.getName(),
                contentIntent.getComponent().getClassName());
        assertEquals(sessionGeneration, contentIntent.getLongExtra(
                CallConstants.EXTRA_SESSION_GENERATION, -1));
        assertNotNull(notification.actions);
        assertEquals(2, notification.actions.length);
        for (Notification.Action action : notification.actions) {
            Intent actionIntent = shadowOf(action.actionIntent).getSavedIntent();
            assertEquals("call-2", actionIntent.getStringExtra(CallConstants.EXTRA_CALL_ID));
            assertEquals(pushGeneration, actionIntent.getLongExtra(
                    CallConstants.EXTRA_CALL_GENERATION, -1));
            assertEquals(sessionGeneration, actionIntent.getLongExtra(
                    CallConstants.EXTRA_SESSION_GENERATION, -1));
        }
    }

    @Test
    public void manifestDeclaresRequiredForegroundServiceTypes() throws Exception {
        PackageManager packageManager = context.getPackageManager();
        ServiceInfo incoming = packageManager.getServiceInfo(
                new ComponentName(context, IncomingCallService.class), 0);
        ServiceInfo inCall = packageManager.getServiceInfo(
                new ComponentName(context, InCallService.class), 0);
        ServiceInfo projection = packageManager.getServiceInfo(
                new ComponentName(context, MediaProjectionCallService.class), 0);

        assertTrue((incoming.getForegroundServiceType()
                & FOREGROUND_SERVICE_TYPE_PHONE_CALL) != 0);
        assertTrue((inCall.getForegroundServiceType()
                & FOREGROUND_SERVICE_TYPE_PHONE_CALL) != 0);
        assertTrue((inCall.getForegroundServiceType()
                & FOREGROUND_SERVICE_TYPE_MICROPHONE) != 0);
        assertTrue((projection.getForegroundServiceType()
                & FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION) != 0);
    }

    @Test
    public void sampleDoesNotDeclareAConnectionKeepingForegroundService() throws Exception {
        PackageInfo packageInfo = context.getPackageManager().getPackageInfo(
                context.getPackageName(), PackageManager.GET_SERVICES);
        Set<String> callScopedForegroundServices = new HashSet<>(Arrays.asList(
                IncomingCallService.class.getName(),
                InCallService.class.getName(),
                MediaProjectionCallService.class.getName()));

        assertNotNull(packageInfo.services);
        for (ServiceInfo service : packageInfo.services) {
            assertFalse((service.getForegroundServiceType()
                    & FOREGROUND_SERVICE_TYPE_SPECIAL_USE) != 0);
            if (service.getForegroundServiceType() != 0) {
                assertTrue(callScopedForegroundServices.contains(service.name));
            }
        }
    }

    @Test
    public void ongoingNotificationCarriesTheOwningSessionGeneration() {
        long generation = 42;

        Notification notification = CallNotificationManager.getInstance(context)
                .buildOngoing("Alice", false, 0, generation);

        assertNotNull(notification.actions);
        assertEquals(1, notification.actions.length);
        Intent hangUpIntent = shadowOf(notification.actions[0].actionIntent).getSavedIntent();
        assertEquals(CallConstants.ACTION_HANG_UP, hangUpIntent.getAction());
        assertEquals(generation, hangUpIntent.getLongExtra(
                CallConstants.EXTRA_SESSION_GENERATION, -1));
        Intent contentIntent = shadowOf(notification.contentIntent).getSavedIntent();
        assertEquals(generation, contentIntent.getLongExtra(
                CallConstants.EXTRA_SESSION_GENERATION, -1));
    }
}

File: Kotlin-OneToOneCallSample/app/src/test/java/com/stringee/kotlin_onetoonecallsample/stringee/integration/CallAndroidIntegrationTest.kt
package com.stringee.kotlin_onetoonecallsample.stringee.integration

import android.content.ComponentName
import android.content.Context
import android.content.pm.PackageManager
import android.content.pm.ServiceInfo
import androidx.test.core.app.ApplicationProvider
import com.stringee.kotlin_onetoonecallsample.activity.MainActivity
import com.stringee.kotlin_onetoonecallsample.stringee.activity.CallActivity
import com.stringee.kotlin_onetoonecallsample.stringee.common.CallConstants
import com.stringee.kotlin_onetoonecallsample.stringee.common.CallNotificationManager.Companion.getInstance
import com.stringee.kotlin_onetoonecallsample.stringee.manager.ActivePushCall
import com.stringee.kotlin_onetoonecallsample.stringee.manager.ActivePushCall.claim
import com.stringee.kotlin_onetoonecallsample.stringee.manager.ActivePushCall.clear
import com.stringee.kotlin_onetoonecallsample.stringee.manager.ActivePushCall.generation
import com.stringee.kotlin_onetoonecallsample.stringee.manager.ActivePushCall.owns
import com.stringee.kotlin_onetoonecallsample.stringee.service.InCallService
import com.stringee.kotlin_onetoonecallsample.stringee.service.IncomingCallService
import com.stringee.kotlin_onetoonecallsample.stringee.service.MediaProjectionCallService
import org.junit.After
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows
import org.robolectric.annotation.Config
import java.util.Arrays

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [35])
class CallAndroidIntegrationTest {
    private val context: Context = ApplicationProvider.getApplicationContext<Context>()

    @After
    fun tearDown() {
        clear()
    }

    @Test
    fun pushIncomingNotificationTargetsMainActivityAndCarriesPushOwnership() {
        Assert.assertTrue(claim("call-1"))
        val generation = generation

        val notification = getInstance(context)
            .buildIncomingFromPush("alice", "Alice", true, "call-1", generation)

        Assert.assertNotNull(notification.contentIntent)
        Assert.assertNotNull(notification.actions)
        Assert.assertEquals(2, notification.actions.size.toLong())
        var foundAnswer = false
        var foundReject = false
        for (action in notification.actions) {
            val pendingIntent = action.actionIntent
            val savedIntent = Shadows.shadowOf(pendingIntent).getSavedIntent()
            if (CallConstants.ACTION_REJECT == savedIntent.getAction()) {
                foundReject = owns(
                    savedIntent.getStringExtra(CallConstants.EXTRA_CALL_ID),
                    savedIntent.getLongExtra(CallConstants.EXTRA_CALL_GENERATION, -1)
                )
            }
            if (ActivePushCall.PendingAction.ANSWER.name == savedIntent.getStringExtra(CallConstants.EXTRA_PENDING_ACTION)) {
                foundAnswer = owns(
                    savedIntent.getStringExtra(CallConstants.EXTRA_CALL_ID),
                    savedIntent.getLongExtra(CallConstants.EXTRA_CALL_GENERATION, -1)
                )
            }
        }
        Assert.assertTrue(foundAnswer)
        Assert.assertTrue(foundReject)
        val contentIntent = Shadows.shadowOf(notification.contentIntent).getSavedIntent()
        Assert.assertEquals(
            MainActivity::class.java.getName(),
            contentIntent.getComponent()!!.getClassName()
        )
    }

    @Test
    fun sdkIncomingNotificationTargetsCallActivityAndCarriesBothOwners() {
        Assert.assertTrue(claim("call-2"))
        val pushGeneration = generation
        val sessionGeneration: Long = 81

        val notification = getInstance(context)
            .buildIncomingFromSdk(
                "bob", "Bob", false, "call-2",
                pushGeneration, sessionGeneration
            )

        val contentIntent = Shadows.shadowOf(notification.contentIntent).getSavedIntent()
        Assert.assertEquals(
            CallActivity::class.java.getName(),
            contentIntent.getComponent()!!.getClassName()
        )
        Assert.assertEquals(
            sessionGeneration, contentIntent.getLongExtra(
                CallConstants.EXTRA_SESSION_GENERATION, -1
            )
        )
        Assert.assertNotNull(notification.actions)
        Assert.assertEquals(2, notification.actions.size.toLong())
        for (action in notification.actions) {
            val actionIntent = Shadows.shadowOf(action.actionIntent).getSavedIntent()
            Assert.assertEquals("call-2", actionIntent.getStringExtra(CallConstants.EXTRA_CALL_ID))
            Assert.assertEquals(
                pushGeneration, actionIntent.getLongExtra(
                    CallConstants.EXTRA_CALL_GENERATION, -1
                )
            )
            Assert.assertEquals(
                sessionGeneration, actionIntent.getLongExtra(
                    CallConstants.EXTRA_SESSION_GENERATION, -1
                )
            )
        }
    }

    @Test
    @Throws(Exception::class)
    fun manifestDeclaresRequiredForegroundServiceTypes() {
        val packageManager = context.getPackageManager()
        val incoming = packageManager.getServiceInfo(
            ComponentName(context, IncomingCallService::class.java), 0
        )
        val inCall = packageManager.getServiceInfo(
            ComponentName(context, InCallService::class.java), 0
        )
        val projection = packageManager.getServiceInfo(
            ComponentName(context, MediaProjectionCallService::class.java), 0
        )

        Assert.assertTrue(
            (incoming.getForegroundServiceType()
                    and ServiceInfo.FOREGROUND_SERVICE_TYPE_PHONE_CALL) != 0
        )
        Assert.assertTrue(
            (inCall.getForegroundServiceType()
                    and ServiceInfo.FOREGROUND_SERVICE_TYPE_PHONE_CALL) != 0
        )
        Assert.assertTrue(
            (inCall.getForegroundServiceType()
                    and ServiceInfo.FOREGROUND_SERVICE_TYPE_MICROPHONE) != 0
        )
        Assert.assertTrue(
            (projection.getForegroundServiceType()
                    and ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION) != 0
        )
    }

    @Test
    @Throws(Exception::class)
    fun sampleDoesNotDeclareAConnectionKeepingForegroundService() {
        val packageInfo = context.getPackageManager().getPackageInfo(
            context.getPackageName(), PackageManager.GET_SERVICES
        )
        val callScopedForegroundServices: MutableSet<String?> = HashSet<String?>(
            Arrays.asList<String?>(
                IncomingCallService::class.java.getName(),
                InCallService::class.java.getName(),
                MediaProjectionCallService::class.java.getName()
            )
        )

        Assert.assertNotNull(packageInfo.services)
        for (service in packageInfo.services!!) {
            Assert.assertFalse(
                (service.getForegroundServiceType()
                        and ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE) != 0
            )
            if (service.getForegroundServiceType() != 0) {
                Assert.assertTrue(callScopedForegroundServices.contains(service.name))
            }
        }
    }

    @Test
    fun ongoingNotificationCarriesTheOwningSessionGeneration() {
        val generation: Long = 42

        val notification = getInstance(context)
            .buildOngoing("Alice", false, 0, generation)

        Assert.assertNotNull(notification.actions)
        Assert.assertEquals(1, notification.actions.size.toLong())
        val hangUpIntent = Shadows.shadowOf(notification.actions[0].actionIntent).getSavedIntent()
        Assert.assertEquals(CallConstants.ACTION_HANG_UP, hangUpIntent.getAction())
        Assert.assertEquals(
            generation, hangUpIntent.getLongExtra(
                CallConstants.EXTRA_SESSION_GENERATION, -1
            )
        )
        val contentIntent = Shadows.shadowOf(notification.contentIntent).getSavedIntent()
        Assert.assertEquals(
            generation, contentIntent.getLongExtra(
                CallConstants.EXTRA_SESSION_GENERATION, -1
            )
        )
    }
}
